File: latest/week3/workshop7_m9/W07_M09_demo.py
"""W07-M09: Audience Adaptation — Python — UNYT"""
import numpy as np, pandas as pd, matplotlib.pyplot as plt, os
os.makedirs("output", exist_ok=True)

nf = pd.read_csv("netflix.csv")
nf["date_added"] = pd.to_datetime(nf["date_added"].str.strip(), errors="coerce")
nf["year_added"] = nf["date_added"].dt.year
yearly = nf.query("2016<=year_added<=2021").groupby(["year_added","type"]).size().reset_index(name="n")
movies = yearly.query("type=='Movie'"); tv = yearly.query("type=='TV Show'")
peak_n = movies.n.max(); latest_n = movies.iloc[-1]["n"]
decline = round((1 - latest_n / peak_n) * 100)

# ═══ THREE AUDIENCE VARIANTS ═══
fig, axes = plt.subplots(1, 3, figsize=(22, 6))

# EXECUTIVE
ax = axes[0]
ax.plot(tv.year_added, tv.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax.text(2021.1, latest_n, " Movies", fontsize=9, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.1, tv.iloc[-1]["n"], " TV", fontsize=8, color="#BBB", va="center")
ax.annotate(f"–{decline}%", xy=(2019.5, (peak_n+latest_n)/2),
    fontsize=14, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935"))
ax.set_xlim(2015.5, 2022.5)
ax.set_title(f"EXECUTIVE\n'Movie additions declined {decline}%'\n1 message, grey+accent, callout",
    fontsize=10, fontweight="bold", color="#1565C0")
ax.set_ylabel("Titles Added")

# TECHNICAL
ax = axes[1]
for t, c in [("Movie", "#1565C0"), ("TV Show", "#E53935")]:
    sub = yearly.query("type==@t")
    ax.plot(sub.year_added, sub.n, "-o", color=c, lw=1.5, markersize=4, label=t)
    for _, row in sub.iterrows():
        ax.text(row.year_added, row.n+15, str(int(row.n)), ha="center", fontsize=5, color=c)
ax.legend(fontsize=7)
ax.set_title("TECHNICAL\n'Netflix Additions by Type (2016–2021)'\nBoth series, all values, legend",
    fontsize=10, fontweight="bold", color="#2E7D32")
ax.set_ylabel("Titles Added")

# PUBLIC
ax = axes[2]
ax.bar(movies.year_added, movies.n, color="#1565C0", width=0.6, alpha=0.8)
for _, row in movies.iterrows():
    ax.text(row.year_added, row.n+15, f"{int(row.n):,}", ha="center",
        fontsize=9, fontweight="bold")
ax.set_title("PUBLIC\n'Netflix added fewer movies each year since 2019'\nSimple bar, big labels, no jargon",
    fontsize=10, fontweight="bold", color="#E65100")
ax.set_ylabel("Movies Added"); ax.set_xlabel("Year")

for ax in axes:
    ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("Same Finding, Three Audiences", fontsize=14, fontweight="bold")
plt.tight_layout(); plt.savefig("output/three_audiences.png", dpi=300); plt.close()

# ═══ INDIVIDUAL VARIANTS ═══
# Executive
fig, ax = plt.subplots(figsize=(10, 5.5))
ax.plot(tv.year_added, tv.n, "-o", color="#DDD", lw=0.8, markersize=3)
ax.plot(movies.year_added, movies.n, "-o", color="#1565C0", lw=2.5, markersize=6)
ax.text(2021.15, latest_n, " Movies", fontsize=10, fontweight="bold", color="#1565C0", va="center")
ax.text(2021.15, tv.iloc[-1]["n"], " TV Shows", fontsize=9, color="#BBB", va="center")
ax.annotate(f"–{decline}% decline\nfrom 2019 peak",
    xy=(2020, (peak_n+latest_n)/2), xytext=(2017.5, latest_n*0.65),
    fontsize=11, fontweight="bold", color="#E53935",
    bbox=dict(boxstyle="round,pad=0.4", facecolor="white", edgecolor="#E53935", lw=1.5),
    arrowprops=dict(arrowstyle="->", connectionstyle="arc3,rad=0.3", color="#E53935", lw=1.5))
ax.set_xlim(2015.5, 2022.8)
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.set_title(f"Movie additions declined {decline}% from their 2019 peak",
    fontsize=14, fontweight="bold")
ax.set_ylabel("Titles Added")
plt.tight_layout(); plt.savefig("output/exec_variant.png", dpi=300); plt.close()

# Technical (monthly with rolling mean)
monthly = (nf.dropna(subset=["date_added"]).query("date_added.dt.year>=2016")
    .set_index("date_added").groupby("type").resample("ME").size()
    .reset_index(name="n"))
fig, ax = plt.subplots(figsize=(10, 5.5))
from matplotlib.dates import DateFormatter
for t, c in [("Movie","#1565C0"),("TV Show","#E53935")]:
    sub = monthly.query("type==@t").sort_values("date_added")
    ax.scatter(sub.date_added, sub.n, s=8, c=c, alpha=0.3)
    roll = sub.n.rolling(3).mean()
    ax.plot(sub.date_added, roll, color=c, lw=1.2, label=t)
    roll_std = sub.n.rolling(3).std()
    ax.fill_between(sub.date_added, (roll-1.96*roll_std).clip(lower=0),
        roll+1.96*roll_std, alpha=0.08, color=c)
ax.legend(fontsize=8)
ax.xaxis.set_major_formatter(DateFormatter("%Y"))
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.set_title("Netflix Monthly Additions by Type (2016–2021)\n"
    "3-month rolling mean ± 1.96σ | Raw points for variance", fontsize=12, fontweight="bold")
ax.set_ylabel("Monthly Titles")
plt.tight_layout(); plt.savefig("output/tech_variant.png", dpi=300); plt.close()

# Public
fig, ax = plt.subplots(figsize=(9, 5))
ax.bar(movies.year_added, movies.n, color="#1565C0", width=0.5, alpha=0.8)
for _, row in movies.iterrows():
    ax.text(row.year_added, row.n+15, f"{int(row.n):,}", ha="center",
        fontsize=10, fontweight="bold", color="#1565C0")
ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
ax.set_title("Netflix is adding fewer movies every year since 2019",
    fontsize=16, fontweight="bold")
ax.set_xlabel("Year", fontsize=12); ax.set_ylabel("")
plt.tight_layout(); plt.savefig("output/public_variant.png", dpi=300); plt.close()

# ═══ e-Car: 3 VARIANTS ═══
ec = pd.read_csv("ecar.csv"); ec.columns=[c.strip().replace("  "," ") for c in ec.columns]
ec["Year"] = pd.to_datetime(ec["Approve Date"], format="%m/%d/%Y", errors="coerce").dt.year
ec["Spread"] = ec["Rate"] - ec["Cost of Funds"]
ec = ec.query("2002<=Year<=2012")
yr = ec.groupby("Year").agg(rate=("Rate","mean"),spread=("Spread","mean")).reset_index()

fig, axes = plt.subplots(1, 3, figsize=(22, 6))
# Exec
ax = axes[0]
ax.axvspan(2007.5,2009.5,alpha=0.04,color="#E53935")
ax.plot(yr.Year, yr.spread, color="#DDD", lw=0.8)
ax.plot(yr.Year, yr.rate, "-o", color="#1565C0", lw=2.5, markersize=5)
ax.annotate("–2pp", xy=(2009, yr.rate.min()), fontsize=14, fontweight="bold",
    color="#E53935", bbox=dict(boxstyle="round,pad=0.3", facecolor="white", edgecolor="#E53935"))
ax.set_title("EXECUTIVE\n'Rates dropped 2pp after 2008'", fontsize=10, fontweight="bold", color="#1565C0")
# Tech
ax = axes[1]
tyr = ec.groupby(["Year","Tier"])["Rate"].mean().reset_index()
for t in sorted(ec.Tier.unique()):
    sub = tyr.query("Tier==@t")
    ax.plot(sub.Year, sub.Rate, "-o", lw=0.8, markersize=3, label=f"Tier {t}")
ax.legend(fontsize=6); ax.axvline(x=2008, ls="--", color="#888", lw=0.5)
ax.set_title("TECHNICAL\n'Rate by Tier (2002–2012)'", fontsize=10, fontweight="bold", color="#2E7D32")
# Public
ax = axes[2]
ax.bar(yr.Year, yr.rate, color="#1565C0", width=0.5, alpha=0.8)
ax.set_title("PUBLIC\n'Car loan rates fell after 2008 crisis'", fontsize=10, fontweight="bold", color="#E65100")
for ax in axes: ax.spines["top"].set_visible(False); ax.spines["right"].set_visible(False)
fig.suptitle("e-Car: Same Finding, Three Audiences", fontsize=14, fontweight="bold")
plt.tight_layout(); plt.savefig("output/ecar_three_audiences.png", dpi=300); plt.close()

print("\nAll W07-M09 Python outputs saved")
print("Netflix: exec_variant, tech_variant, public_variant, three_audiences")
print("e-Car: ecar_three_audiences")

File: latest/week3/workshop7_m9/W07_M09_demo.R
# ============================================================
# Workshop 7 — Module 9: Audience Adaptation
# Integrated R Script — UNYT Tirana
# ============================================================
# Same finding → 3 audience variants (executive, technical, public)
library(tidyverse); library(patchwork); dir.create("output", showWarnings = FALSE)

theme_story <- theme_minimal(base_size = 12) +
  theme(panel.grid = element_blank(),
    axis.line = element_line(color = "#EEE", linewidth = 0.3),
    plot.title = element_text(face = "bold", size = 14, lineheight = 1.2),
    plot.subtitle = element_text(color = "#666", size = 10),
    legend.position = "none")
theme_set(theme_story)

# ── LOAD DATA ────────────────────────────────────────────
nf <- read_csv("netflix.csv") |>
  mutate(date_added = mdy(str_trim(date_added)), year_added = year(date_added))
yearly <- nf |> filter(!is.na(year_added), year_added >= 2016, year_added <= 2021) |>
  count(year_added, type)
movies <- yearly |> filter(type == "Movie")
tv <- yearly |> filter(type == "TV Show")
peak <- movies |> slice_max(n, n = 1)
decline <- round((1 - movies$n[movies$year_added == 2021] / peak$n) * 100)

# ═══════════════════════════════════════════════════════
# VARIANT 1: EXECUTIVE
# One message, one chart, one number
# ═══════════════════════════════════════════════════════
p_exec <- ggplot(yearly, aes(x = year_added, y = n, group = type)) +
  geom_line(color = "#DDDDDD", linewidth = 0.8) +
  geom_line(data = movies, color = "#1565C0", linewidth = 2.5) +
  geom_point(data = movies, color = "#1565C0", size = 3) +
  annotate("label", x = 2018.5, y = movies$n[movies$year_added == 2021] * 0.80,
    label = paste0("–", decline, "%\nfrom 2019 peak"),
    size = 4, fontface = "bold", color = "#E53935",
    fill = "white", label.size = 0.5) +
  geom_text(data = yearly |> filter(year_added == 2021),
    aes(label = type, color = type == "Movie"),
    hjust = -0.1, size = 3.5, fontface = "bold", show.legend = FALSE) +
  scale_color_manual(values = c("TRUE" = "#1565C0", "FALSE" = "#BBB")) +
  coord_cartesian(xlim = c(2016, 2022.5)) +
  labs(title = paste0("Movie additions declined ", decline, "% from their 2019 peak"),
       subtitle = "Recommendation: reallocate content budget to originals + local markets",
       x = NULL, y = "Titles Added")
ggsave("output/exec_variant.png", p_exec, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# VARIANT 2: TECHNICAL (Data Science Team)
# Full detail, both series, CI, methodology visible
# ═══════════════════════════════════════════════════════
# Compute monthly with CI
monthly <- nf |>
  filter(!is.na(date_added), year(date_added) >= 2016) |>
  mutate(month = floor_date(date_added, "month")) |>
  count(month, type) |>
  group_by(type) |> arrange(month) |>
  mutate(roll_mean = zoo::rollmean(n, k = 3, fill = NA, align = "right"),
    roll_sd = zoo::rollapply(n, width = 3, FUN = sd, fill = NA, align = "right")) |>
  ungroup()

p_tech <- ggplot(monthly, aes(x = month, y = n, color = type)) +
  geom_point(size = 0.8, alpha = 0.3) +
  geom_line(aes(y = roll_mean), linewidth = 1) +
  geom_ribbon(aes(ymin = pmax(roll_mean - 1.96 * roll_sd, 0),
    ymax = roll_mean + 1.96 * roll_sd, fill = type),
    alpha = 0.1, color = NA) +
  scale_color_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  scale_fill_manual(values = c(Movie = "#1565C0", "TV Show" = "#E53935")) +
  theme_minimal(base_size = 10) +
  theme(panel.grid.major.y = element_line(color = "#F0F0F0"),
    legend.position = "bottom") +
  labs(title = "Netflix Monthly Additions by Type (2016–2021)",
       subtitle = "3-month rolling mean ± 1.96σ ribbon | Raw points shown for variance assessment",
       x = NULL, y = "Monthly Titles Added", color = NULL, fill = NULL,
       caption = "Method: rolling window k=3, align=right | Source: netflix.csv")
ggsave("output/tech_variant.png", p_tech, width = 10, height = 5.5, dpi = 300)

# ═══════════════════════════════════════════════════════
# VARIANT 3: PUBLIC (Blog / Social Media)
# Simplest possible, no jargon, big labels
# ═══════════════════════════════════════════════════════
p_public <- ggplot(movies, aes(x = factor(year_added), y = n)) +
  geom_col(fill = "#1565C0", width = 0.5, alpha = 0.8) +
  geom_text(aes(label = format(n, big.mark = ",")),
    vjust = -0.5, size = 4, fontface = "bold", color = "#1565C0") +
  theme_minimal(base_size = 14) +
  theme(panel.grid = element_blank(),
    axis.line.x = element_line(color = "#EEE"),
    plot.title = element_text(face = "bold", size = 16)) +
  labs(title = "Netflix is adding fewer movies every year since 2019",
       subtitle = "The streaming giant's movie catalog is shrinking",
       x = "Year", y = NULL)
ggsave("output/public_variant.png", p_public, width = 9, height = 5, dpi = 300)

# ═══════════════════════════════════════════════════════
# SIDE-BY-SIDE COMPARISON
# ═══════════════════════════════════════════════════════
ggsave("output/three_audiences.png",
  (p_exec + labs(title = "EXECUTIVE\n1 message, grey+accent, callout")) |
  (p_tech + labs(title = "TECHNICAL\nBoth series, CI ribbon, raw data")) |
  (p_public + labs(title = "PUBLIC\nSimple bar, big labels, no jargon")),
  width = 22, height = 6, dpi = 300)

# ═══════════════════════════════════════════════════════
# e-Car: 3 AUDIENCE VARIANTS
# ═══════════════════════════════════════════════════════
ec <- read_csv("ecar.csv") |>
  mutate(Year = year(mdy(`Approve Date`)),
    Spread = Rate - `Cost of Funds`) |>
  filter(Year >= 2002, Year <= 2012)
yr_ec <- ec |> group_by(Year) |>
  summarise(rate = mean(Rate), spread = mean(Spread))

# Executive
p_ec_exec <- ggplot(yr_ec) +
  annotate("rect", xmin = 2007.5, xmax = 2009.5, ymin = -Inf, ymax = Inf,
    fill = "#E53935", alpha = 0.04) +
  geom_line(aes(x = Year, y = spread), color = "#DDD", linewidth = 0.8) +
  geom_line(aes(x = Year, y = rate), color = "#1565C0", linewidth = 2.5) +
  annotate("label", x = 2005, y = min(yr_ec$rate) * 0.92,
    label = "–2pp\nrate drop", size = 4, fontface = "bold",
    color = "#E53935", fill = "white", label.size = 0.5) +
  annotate("text", x = 2012.3, y = yr_ec$rate[yr_ec$Year == 2012],
    label = "Rate", fontface = "bold", color = "#1565C0", size = 3.5, hjust = 0) +
  scale_x_continuous(limits = c(2002, 2013)) +
  labs(title = "Loan rates dropped 2pp after the 2008 crisis",
       subtitle = "Action: integrate CI-based early warning into pricing reviews")

# Technical
tier_yr <- ec |> group_by(Year, Tier) |> summarise(rate = mean(Rate), .groups = "drop")
p_ec_tech <- ggplot(tier_yr, aes(x = Year, y = rate, color = factor(Tier))) +
  geom_line(linewidth = 0.8) + geom_point(size = 1.5) +
  geom_vline(xintercept = 2008, linetype = "dashed", color = "#888") +
  theme_minimal(base_size = 9) +
  theme(legend.position = "bottom") +
  labs(title = "Mean Rate by Credit Tier (2002–2012)",
       subtitle = "Tier-stratified view | Vertical line = 2008 crisis onset",
       color = "Credit Tier", y = "Mean Rate (%)")

# Public
p_ec_public <- ggplot(yr_ec, aes(x = Year, y = rate)) +
  geom_col(fill = "#1565C0", width = 0.5, alpha = 0.8) +
  theme_minimal(base_size = 14) +
  theme(panel.grid = element_blank()) +
  labs(title = "Car loan rates dropped sharply\nafter the 2008 financial crisis",
       subtitle = "Average interest rate on auto loans, by year", y = "Rate (%)")

ggsave("output/ecar_three_audiences.png",
  (p_ec_exec + labs(title = "EXECUTIVE")) |
  (p_ec_tech + labs(title = "TECHNICAL")) |
  (p_ec_public + labs(title = "PUBLIC")),
  width = 22, height = 6, dpi = 300)

# ═══════════════════════════════════════════════════════
# MEDIUM ADAPTATION: same exec chart, 3 channels
# ═══════════════════════════════════════════════════════
# Presentation: large fonts, minimal
p_present <- p_exec +
  theme(text = element_text(size = 16),
    plot.title = element_text(size = 20)) +
  labs(title = "FOR PROJECTION\n(large fonts, minimal detail)")

# Email: self-contained with caption
p_email <- p_exec +
  labs(title = "FOR EMAIL\n(self-contained: title + chart tell full story)",
       caption = "Data: Netflix dataset. Analysis: Data Viz Team. Date: March 2025.")

# Social: bold, crop-friendly
p_social <- p_public +
  theme(text = element_text(size = 18),
    plot.title = element_text(size = 22),
    plot.margin = margin(20, 20, 20, 20)) +
  labs(title = "FOR SOCIAL MEDIA\n(big, bold, 3-second read)")

ggsave("output/medium_adaptation.png",
  p_present / p_email / p_social,
  width = 10, height = 16, dpi = 300)

cat("\n── All W07-M09 R outputs saved ──\n")
cat("Variants: exec, tech, public (Netflix + e-Car)\n")
cat("Medium: presentation, email, social\n")
